library/src/blas1/rocblas_copy.hpp

```diff
@@ -50,17 +53,31 @@ rocblas_status rocblas_copy_template(rocblas_handle handle,
     if(n <= 0 || !batch_count)
         return rocblas_status_success;
 
+    if(!x || !y)
+        return rocblas_status_invalid_pointer;
+
+    if(batch_count < 0)
+        return rocblas_status_invalid_size;
```

Comment:
Although it is redundant to do error checking inside the `*_template` routine, the argument error checks for `COPY` are minor, and should not affect performance.

The convention of not doing error-checking in `*_template` routines is not a **hard rule**, but a guideline. It is designed this way so as to reduce unnecessary overhead. On the other hand, not doing device memory allocation and logging in `*_template` routines, are hard rules.

We must check for `n <=0 || !batch_count` and return success, _before_ we check other arguments for errors, because the success short-circuit takes precedence. We must also do this early success exit in the `*_template` routine.

So to avoid error checks in `*_template` routine would require splitting the `*_template` function into two functions, one for rocBLAS and one for Solver, or adding another `bool` template argument to enable or disable the error checks. Rather than unnecessarily complicate the code, we will leave the error checks in `*_template`, since they take almost no CPU time (less than 5 cycles).